lighthouse-logger/index.js

```diff
@@ -118,33 +166,153 @@ class Log {
+   * @return {(this: T, ...args: Args) => R}
+   */
+  static timeDecorate(originalFn, opts) {
+    /**
+     * @type {(this: *, ...args: *[]) => string}
+     */
+    const computeMsg = (_this, args) => {
+      if (typeof opts.msg === 'string') return opts.msg;
+      if (typeof opts.msg === 'function') return opts.msg.apply(_this, args);
+      throw new Error('expected msg');
+    };
+
+    /**
+     * @type {(this: *, ...args: *[]) => string}
+     */
+    const computeId = (_this, args) => {
+      if (typeof opts.id === 'string') return opts.id;
+      if (typeof opts.id === 'function') return opts.id.apply(_this, args);
+      return `lh:${originalFn.name}`;
+    };
+
+    /**
+     * @type {(this: T, ...args: Args) => R}
+     */
+    const fn = function timeDecoratedFn(...args) {
+      const timeStartLogLevel = opts.timeStartLogLevel || 'log';
+      const timeEndLogLevel = opts.timeEndLogLevel || 'verbose';
+
+      const status = {msg: computeMsg(this, args), id: computeId(this, args)};
+      Log.time(status, timeStartLogLevel);
+
+      let result;
+      try {
+        result = originalFn.apply(this, args);
+      } catch (err) {
+        Log.timeEnd(status, timeEndLogLevel);
+        // intercept any errors and elide the time decoration from the stack trace
+        err.stack = err.stack.replace(/.* at Function\.timeDecoratedFn .*\n/g, '');
+        throw err;
+      }
+
+      if (result && typeof result.then === 'function') {
+        return result.then((/** @type {any} */ value) => {
+          Log.timeEnd(status, timeEndLogLevel);
+          return value;
+        }).catch((/** @type {any} */ err) => {
+          Log.timeEnd(status, timeEndLogLevel);
+          // intercept any errors and elide the time decoration from the stack trace
+          err.stack = err.stack.replace(/.* at Function\.timeDecoratedFn .*\n/, '');
+          throw err;
+        });
+      } else {
+        Log.timeEnd(status, timeEndLogLevel);
+        return result;
+      }
+    };
+    return fn;
   }
+  /* eslint-enable no-invalid-this */
 
-  static warn(title, ...args) {
-    Log.events.issueWarning(title, args);
-    return Log._logToStdErr(`${title}:warn`, args);
+  /**
+   * Decorates (like timeDecorate) specified methods of a class.
+   * If decorating instance methods, use the class's prototype.
+   * If decorating static methods, use the class directly.
+   * @template {Object|Function} Class
+   * @template {keyof Class} Prop
+   * @param {Class} klass
+   * @param {{[key in Prop]: TimeDecorateOpts<Class, Parameters<IsFunction<Class[key]>>>}} methods
+   */
+  static timeDecorateClass(klass, methods) {
+    for (const [method, opts] of Object.entries(methods)) {
+      if (!opts.id) {
+        const className = (typeof klass === 'function' ? klass : klass.constructor).name;
+        opts.id = `lh:${className}:${method}`;
+      }
+      /** @type {IsFunction<Class[typeof method]>} */
+      const original = klass[method];
```

Comment:
I'm concerned that basically everything in this and the above function are typed as `any` :)

I wonder if there's something simpler (or defining more in a `d.ts` file if necessary) to get us *some* constraints enforced. Or we go the opposite, keeping the interface as strict as possible and going explicitly `any` inside (this is partially where `timeDecorate` has gone, at least with the return value).

Reply:
I'm not sure. My main goal was to only allow property names of functions of `klass` in `methods`. The current typing achieves that, but the local typing is a bit out of my reach (or TypeScript's?). But the ol' gut type checker seems right to me (AKA it _looks_ correct).